### PR TITLE
Implementation of RFC-3986 algorithm for the URL resolve

### PR DESCRIPTION
**Pull request by TechAlchemistry**
Improve of algorithm the URL Resolve. Now is based on the [RFC-3986 section 5.2](http://tools.ietf.org/html/rfc3986#section-5.2).

### CODE REVIEW

**Comment by TechAlchemistry** [comment]
Why failed? I checked it before create commit. Maybe it because I changed the test suit? 

How I can see why it failed?

**Comment by gsouf** [comment]
Click on the detail link: https://travis-ci.org/serp-spider/core/jobs/134313100#L245

About the commit, the method `removePathDotSegments` is not desired here. The goal of resolve is to offer the new url as it is.

For instance : 
- resolving `http://foo.bar/foo`  and  `../bar` should result in  `http://foo.bar/foo/../bar`
- resolving `http://foo.bar/foo`  and  `http://foo.bar/foo/../bar` should result in  `http://foo.bar/foo/../bar`

`resolve` should not transform the url. This would be the job of some `realpath`-like method.

File: src/Core/Url/QueryParam.php [hunk 1]
```diff
@@ -72,11 +72,15 @@ public function isRaw()
...
+     * @return string the parameter on this format: ``name=value`` or ``name``
      */
     public function generate()
     {
-        return $this->getName() . '=' . $this->getValue();
+        $result = $this->getName();
+        if (!is_null($this->value)) {
+            $result .= '=' . $this->getValue();
+        }
+        return $result;
```
Comment (gsouf) [inline src/Core/Url/QueryParam.php:83]: For ease of reviewing you should place this change in its own PR (or at least it's own commit in the PR), mixing things when reviewing is hard for maintainers and bad for history. The commit is about how resolve url works, and this is about how url parameters are rendered

File: test/suites/TDD/Core/UrlTest.php [hunk 2]
```diff
@@ -164,4 +164,64 @@ public function testResolveAsBadClass()
         $this->setExpectedException(\InvalidArgumentException::class);
         $url->resolve('//bar', Cookie::class);
     }
+
+    /**
+     * @dataProvider RFC3986ResolveDataProvider
```
Comment (gsouf) [inline test/suites/TDD/Core/UrlTest.php:169]: Didn't know dataProvider until know, good to know

**Comment by TechAlchemistry** [comment]
I wrote resolve method based on RFC. If you think it not right, can you fix it?

(And probably need to create the issue in RFC? Or I misunderstood it?)

File: test/suites/TDD/Core/UrlTest.php (hunk 2 above)
Comment (TechAlchemistry) [inline test/suites/TDD/Core/UrlTest.php:169]: You are welcome ;)

File: src/Core/Url/QueryParam.php (hunk 1 above)
Comment (TechAlchemistry) [inline src/Core/Url/QueryParam.php:83]: How I can split the exists commit on two?

Comment (gsouf) [inline src/Core/Url/QueryParam.php:83]: soft reset, make your 2 commits, and push with force. But for this time it's ok, and the change is minor !

Comment (TechAlchemistry) [inline src/Core/Url/QueryParam.php:83]: Do I understand: I do not need to do that?

**Comment by gsouf** [comment]
I'm still thinking. Though it is not common, it can be confusing in extreme cases to remove dots (`./` and `../`), but in most of cases it wont give a difference. I can't imagine a case it would be a problem, but in some case you might want to keep the full path (with dots)

Well I'm thinking of adding a third parameter that would control simplification of the path (removing dot links) and that would be enabled by default .

**Comment by TechAlchemistry** [comment]
I think it will be a excess, but if you think it necessary, I can do it.

**Comment by gsouf** [comment]
What I'm talking about is the optional normalization as described here: https://tools.ietf.org/html/rfc3986#section-6.2.2

**Comment by TechAlchemistry** [comment]
Or we can add this option if it is really necessary?

**Comment by gsouf** [comment]
Well, you are right, I dont like the idea of a third parameter as well.

I think it could be very helpful, for url comparison, to have an additional method (`normalize`) that would do that. 

In a regular workflow, if you want to resolve and normalize an url, that would give:

``` php
$resolvedAndNormalizedUrl = $url->resolve("someUrl")->normalize();
```

In this case resolve would never normalize the url, and things remains clear for everyone

Opening an issue for that.

**Comment by gsouf** [comment]
I'm closing the PR because it's done differently in #22 I took a few inputs from your PR, thanks for your contribution

**Pull request closed without merging**